src/bin/ion/commands/beta/generate/generator.rs

```diff
@@ -386,14 +423,95 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
     /// ```
     /// For the above schema, both `fields` and `type` constraints map to different abstract data types
     /// respectively Struct(with given fields `source` and `destination`) and Value(with a single field that has String data type).
-    fn verify_abstract_data_type_consistency(
+    fn verify_and_update_abstract_data_type(
         &mut self,
         current_abstract_data_type: AbstractDataType,
+        tera_fields: &mut Vec<Field>,
         code_gen_context: &mut CodeGenContext,
     ) -> CodeGenResult<()> {
         if let Some(abstract_data_type) = &code_gen_context.abstract_data_type {
-            if abstract_data_type != &current_abstract_data_type {
-                return invalid_abstract_data_type_error(format!("Can not determine abstract data type as current constraint {} conflicts with prior constraints for {}.", current_abstract_data_type, abstract_data_type));
+            match abstract_data_type {
+                // In the case when a `type` constraint occurs before `element` constraint. The element type for the sequence
+                // needs to be updated based on `element` constraint whereas sequence type will be used as per `type` constraint.
+                // e.g. For a schema as below:
+                // ```
+                // type::{
+                //   name: sequence_type,
+                //   type: sexp,
+                //   element: string,
+                // }
+                // ```
+                // Here, first `type` constraint would set the `AbstractDataType::Sequence{ element_type: T, sequence_type: "sexp"}`
+                // which uses generic type T and sequence type is sexp. Next `element` constraint would
+                // set the `AbstractDataType::Sequence{ element_type: String, sequence_type: "list"}`.
+                // Now this method performs verification that if the above described case occurs
+                // then it updates the `element_type` as per `element` constraint
+                // and `sequence_type` as per `type` constraint.
+                AbstractDataType::Sequence {
+                    element_type,
+                    sequence_type,
+                } if abstract_data_type != &current_abstract_data_type
+                    && (element_type == "Object" || element_type == "T")
```

Comment:
These constants should be given a name rather than having "magic values" in several places.

Reply:
Sure. I will make constants for these values.

Reply:
This has been changed to use `None` instead.